src/hotspot/share/classfile/dictionary.cpp

```diff
@@ -132,12 +132,17 @@ bool Dictionary::resize_if_needed() {
   return (desired_size != 0);
 }
 
+bool DictionaryEntry::is_valid_protection_domain(Handle protection_domain) {
+
+  return protection_domain() == NULL || !java_lang_System::allow_security_manager()
+        ? true
+        : contains_protection_domain(protection_domain());
+}
+
 bool DictionaryEntry::contains_protection_domain(oop protection_domain) const {
-  // Lock the pd_set list.  This lock cannot safepoint since the caller holds
-  // a Dictionary entry, which can be moved if the Dictionary is resized.
-  MutexLocker ml(ProtectionDomainSet_lock, Mutex::_no_safepoint_check_flag);
 #ifdef ASSERT
   if (protection_domain == instance_klass()->protection_domain()) {
+    MutexLocker ml(ProtectionDomainSet_lock, Mutex::_no_safepoint_check_flag);
```

Comment:
By splitting the lock scope into two blocks you have lost the atomicity of the entire action in a debug build, and now you check a potentially different pd_set().

Reply:
If the dictionary entry's class matches the protection domain, then it should never be added to the pd_set list.  So it doesn't need to be locked to check that.  It doesn't need atomicity.

Reply:
I missed the re-check at line 163 means we are going to return immediately after we execute the debug code.